### ISSUE TITLE
Segmentation fault [Debian Testing]

### ISSUE THREAD

**Issue by criztovyl** [Unode/firefox_decrypt#9]
```
$ git describe --tags
0.5.4-3-gf6d16af
$ ./firefox_decrypt -vvvvvvvvv #  verbose overflow :D
2017-02-03 15:05:28,862 - DEBUG - Parsed commandline arguments: Namespace(choice=None, export_pass=False, list=False, no_interactive=False, profile='~/.mozilla/firefox', verbose=9)
2017-02-03 15:05:28,862 - DEBUG - Skipping password store test, not exporting
2017-02-03 15:05:28,862 - DEBUG - Loading NSS library from libnss3.so
2017-02-03 15:05:28,864 - DEBUG - Reading profiles from /home/christoph/.mozilla/firefox/profiles.ini
2017-02-03 15:05:28,865 - DEBUG - Read profiles ['General', 'Profile0', 'Profile1', 'Profile2', 'Profile3', 'Profile4']
Select the Firefox profile you wish to decrypt
1 -> 91u1xh15.[Profile 1]
2 -> vu3dnx5b.[Profile 2]
3 -> 1d8vcooi.[Profile 3]
4 -> emof2pes.[Profile 4]
5 -> 8a52xmxd.Fresh
Choice: 5                         
2017-02-03 15:05:47,516 - DEBUG - Profile selection matched 8a52xmxd.Fresh

Master Password for profile /home/christoph/.mozilla/firefox/8a52xmxd.Fresh: 
2017-02-03 15:05:53,011 - DEBUG - Initializing NSS with profile path '/home/christoph/.mozilla/firefox/8a52xmxd.Fresh'
2017-02-03 15:05:53,016 - DEBUG - Initializing NSS returned 0
2017-02-03 15:05:53,016 - DEBUG - Retrieving internal key slot
2017-02-03 15:05:53,016 - DEBUG - Internal key slot -1009786480
2017-02-03 15:05:53,017 - DEBUG - Authenticating with password '[Password]'
Segmentation fault
```
Removed Profile 1-4 Names & the Password from Log.

- It's failing at [Line 275](https://github.com/Unode/firefox_decrypt/blob/master/firefox_decrypt.py#L275) (otherwise the debug message from 276 would be printed ^^)
- both Python 2 and 3
- libnss3 2:3.26.2-1
- f6d16af1594e8f1716fc7a4f89fde9fa5ba7ed9f

**Comment by Unode** [comment]
Hi Christoph, thanks for the detailed bug report.

I'll try to reproduce the problem when possible. Hopefully not another #2.
In the meantime, can you try compiling [the same version of NSS from source](https://ftp.mozilla.org/pub/security/nss/releases/NSS_3_26_2_RTM/src/) and pointing the script to the corresponding `libnss3.so`?

Also does changing your master password to something containing only ASCII characters still reproduce the segfault?

**Comment by criztovyl** [comment]
My password does contain only ASCII (w/ spaces, don't know if it's relevant)

**Comment by Unode** [comment]
Space should be fine, as should other characters, just trying to isolate variables.

**Comment by criztovyl** [comment]
Is there a list of required libraries for building nss?

**Comment by criztovyl** [comment]
nvm, was just libc6-dev-i386. Little bit strange b/c I'm on amd64...

**Comment by criztovyl** [comment]
```
$ firefox_decrypt -vvv
2017-02-03 17:22:53,256 - DEBUG - Parsed commandline arguments: Namespace(choice=None, export_pass=False, list=False, no_interactive=False, profile='~/.mozilla/firefox', verbose=3)
2017-02-03 17:22:53,257 - DEBUG - Skipping password store test, not exporting
Traceback (most recent call last):
  File "/home/christoph/.local/opt/bin/firefox_decrypt", line 720, in <module>
    main()
  File "/home/christoph/.local/opt/bin/firefox_decrypt", line 704, in main
    nss = NSSInteraction()
  File "/home/christoph/.local/opt/bin/firefox_decrypt", line 175, in __init__
    self.NSS = CDLL("/home/christoph/build/nss-3.26.2/nss/lib/nss/Linux4.8_x86_cc_glibc_PTH_DBG.OBJ/libnss3.so")
  File "/usr/lib/python2.7/ctypes/__init__.py", line 362, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: /home/christoph/build/nss-3.26.2/nss/
  lib/nss/Linux4.8_x86_cc_glibc_PTH_DBG.OBJ/libnss3.so: wrong ELF class: ELFCLASS32
```
(I wrapped the OSError line)

Yeah. It does not make sense to have 32bit lib here. Did I built wrong? Or is sth broken in testing?
Will try again.

**Comment by Unode** [comment]
I remember bumping into this issue too. There's some environment flag that needs to be set for it to compile as 64bits. Let me see if I can find that.

**Comment by criztovyl** [comment]
`-DUSE_64` maybe? (Makefile L73ff)

**Comment by Unode** [comment]
I'm also trying to compile it... haven't done it in a while and there's something off with the instructions. Not so straightforward... I'll let you know if I succeed but I'm pretty sure `USE_64=1 make ...` is the right way to go about it.

**Comment by Unode** [comment]
So:
```
tar xf nss-3.26.2-with-nspr-4.12.tar.gz
cd nss-3.26.2/nss/
USE_64=1 gmake nss_build_all
ls ../dist/Linux4.6_x86_64_gcc_glibc_PTH_64_DBG.OBJ/lib/libnss3.so
```
does it for me.

**Comment by criztovyl** [comment]
I'm also done building successfully. But:

I changed [line 175](https://github.com/Unode/firefox_decrypt/blob/master/firefox_decrypt.py#L175) to:
```
self.NSS = CDLL("/home/christoph/build/nss-3.26.2/nss/lib/nss/Linux4.8_x86_64_cc_glibc_PTH_64_DBG.OBJ/libnss3.so")
#self.load_libnss()

```
 and put the compiled directory containing the libnss3.so to `LD_LIBRARY_PATH`.

Still get `Segmenatation fault`.
Or is that the wrong way? :D

**Comment by Unode** [comment]
No need to add to `LD_LIBRARY_PATH`. Actually if you have LD_LIBRARY_PATH set you should `unset LD_LIBRARY_PATH`.

Just changing the code to point to the `libnss3.so` should be sufficient.

**Comment by criztovyl** [comment]
```
2017-02-03 18:26:17,973 - DEBUG - Initializing NSS returned -1
2017-02-03 18:26:17,974 - ERROR - Couldn't initialize NSS, maybe '/home/christoph/.mozilla/firefox/8a52xmxd.Fresh' is not a valid profile?
2017-02-03 18:26:17,974 - DEBUG - Error during a call to NSS library, trying to obtain error info
2017-02-03 18:26:17,974 - DEBUG - PR_LOAD_LIBRARY_ERROR: Failure to load dynamic library
```

**Comment by Unode** [comment]
That's an odd one... does it also happen with other profiles?
I can't reproduce it locally with Python 2 nor Python 3.

**Comment by criztovyl** [comment]
Happens with all my profiles, with compiled lib and system lib.

**Comment by Unode** [comment]
Hum, then it's likely another lib causing problems.

Can you run the following on the version you just compiled:
```
% ldd libnss3.so 
	linux-vdso.so.1 (0x00007ffc435a8000)
	libnssutil3.so => /usr/lib64/libnssutil3.so (0x00007fd28cce3000)
	libplc4.so => /usr/lib64/libplc4.so (0x00007fd28cade000)
	libplds4.so => /usr/lib64/libplds4.so (0x00007fd28c8d9000)
	libnspr4.so => /usr/lib64/libnspr4.so (0x00007fd28c69b000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fd28c47e000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fd28c279000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fd28bec9000)
	librt.so.1 => /lib64/librt.so.1 (0x00007fd28bcc1000)
	/lib64/ld-linux-x86-64.so.2 (0x000055aa6fbc5000)
```

I'll have to get a debian testing machine to try to reproduce it although I'm afraid this might be a dead end or something I can't control from the script.

**Comment by criztovyl** [comment]
```
$ ldd libnss3.so 
	linux-vdso.so.1 (0x00007ffc131db000)
	libnssutil3.so => /usr/lib/x86_64-linux-gnu/libnssutil3.so (0x00007f0a086ae000)
	libplc4.so => /usr/lib/x86_64-linux-gnu/libplc4.so (0x00007f0a084a9000)
	libplds4.so => /usr/lib/x86_64-linux-gnu/libplds4.so (0x00007f0a082a5000)
	libnspr4.so => /usr/lib/x86_64-linux-gnu/libnspr4.so (0x00007f0a08065000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f0a07e48000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f0a07c42000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f0a078a4000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f0a0769c000)
	/lib64/ld-linux-x86-64.so.2 (0x000055bd42fcb000)
```

**Comment by criztovyl** [comment]
> Happens with all my profiles, with compiled lib and system lib.

Sorry, that was misleading. I meant with system lib all profiles have `Segmentation fault`, with compiled lib that `PR_LOAD_LIBRARY_ERROR`.

**Comment by Unode** [comment]
And `env | grep LIBRARY` doesn't show `LD_LIBRARY_PATH` right?

**Comment by criztovyl** [comment]
Jup, no `LD_LIBRARY_PATH`.

**Comment by Unode** [comment]
I'm out of ideas... Could you perhaps try the/a most/more recent libnss and see if it helps?

**Comment by criztovyl** [comment]
I asked at #debian-next, maybe someone there can reproduce. If not, mybe my setup is simply broken :D
Thank you for your effort :+1:

**Comment by Unode** [comment]
@criztovyl I noticed a pattern which still don't know if it's relevant or not.
It appears that when "keyslot" is represented as a negative number, a segfault is seen. This seems to be the case with #10 and possibly #2 but I can't confirm the last one.

If you have any working profile, can you confirm if this is consistent?

**Comment by Unode** [comment]
@criztovyl A shot in the dark, can you paste the output of: `uname -a`?

**Comment by criztovyl** [comment]
What do you mean by "working profile"? :)
The profile at least is not broken, FF still works with it (it was running last time, I'll check whether no running FF changes sth).

Anyway, unfortunately I can't confirm your pattern, I'm getting seemingly random key slot numbers, both negative and positive.

Also:
```
$ uname -a
Linux chschulz05 4.8.0-2-amd64 #1 SMP Debian 4.8.15-2 (2017-01-04) x86_64 GNU/Linux
```

**Comment by Unode** [comment]
By working profile I mean one that you can read passwords from.
But if the pattern doesn't hold even on the segfault one then it's not it.

I'll work on reproducing it on my side.

**Comment by criztovyl** [comment]
[A wild gdb log appears](https://gist.github.com/criztovyl/8e7baa0aa3efdeef6992198246474600)

**Comment by Unode** [comment]
@criztovyl Thanks for the gdb log. I was expecting something like this but still clueless about the causes.
In the meantime I've found out that there are now [python bindings to libnss](https://pypi.python.org/pypi/python-nss/1.0.0). I managed to initialize it on Debian testing without segfaulting, so I'm quite sure I'm doing something wrong (or not doing it at all) when interacting directly with the NSS API.
I'm planning to give the bindings a try over the next few days... once I find which of the hundreds of methods I have to call :). I'll put up my progress on another branch once I have something.
